README.md
[![Python 3.8](https://github.com/PageotD/juliaset/actions/workflows/python3-8.yaml/badge.svg?branch=develop)](https://github.com/PageotD/juliaset/actions/workflows/python3-8.yaml)

![BSD 3-Clause License](https://img.shields.io/badge/%20License%20-BSD%203--Clause-informational)

# juliaset
A simple random Julia set generator. Can take some time to run, so be patient !

This is a small personal project I did to apply some of the skills I learned during my DevOps career transition training.

This package is not initially intended to be distributed and is therefore not present on the Python Package Index ([PyPi](https://pypi.org/)). However, if you wish, you can clone it, fork it, modify it etc. within the limits of the license.

## Installation

Clone the repository:
```bash
https://github.com/PageotD/juliaset.git
```

Install julia using `pip`:
```bash
pip install ./juliaset --use-feature=in-tree-build
```

Or using `poetry`:
```bash
poetry install ./juliaset
```

## Some examples

```python
# Import julia from the juliaset module
from juliaset import julia

# Create a new set with all default parameters
newset1 = julia()

# Run the newset. It should result in the creation of a new image file
# named newset1.png
newset1.run(fname='newset1')
```

![newset1](https://raw.githubusercontent.com/PageotD/juliaset/develop/docs/images/juilaset-output-example.png)

```python
# Import julia from the juliaset module
from juliaset import julia

# Create a new set with all default parameters
newset2= julia()

# Create a new set with some parameters
# size = 512 will generate a 512x512 pixel image
# norm = False means no normalization will be performed before plotting
newset2 = julia(mirror=True)

# Run the newset. It should result in the creation of a new image file
# named newset2.png
newset2.run(fname='newset2')
```

![newset2](https://raw.githubusercontent.com/PageotD/juliaset/develop/docs/images/juilaset-output-mirror-example.png)

## Parameters

- `size`: size of the array in height and width (not of the output image), default `size=256`
- `dpi`: _dot per inch_ for the output image, default `dpi=300`
- `norm`: normalize the julia set by its maximum value, default `norm=True`
- `mirror`: mirroring the julia set horizontally and vertically, default `mirror=False`
- `escrad`: escape radius, default `escrad=3`
- `niter`: maximum number of iterations, default `niter=250`

## Useful links

- https://en.wikipedia.org/wiki/Julia_set


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [unreleased] - yyyy-mm-dd

<<<<<<< HEAD
=======
### Changed

- `README.md`

>>>>>>> a5a92a9 (Improving README.md)
## [0.3.0] - 2021-08-03

### Added

- `param` method to `JuliaSet` class
- `Dockerfile` for testing package installation
- `setup.py`
- `tests` folder
- `juliaset` folder with `__init__.py` file
- `MANIFEST.in`
- `pyproject.toml`
- `tests/test_class_juliaset.py`
- `tests/tests_julia.py`

### Changed

- remove arguments from `__init__(self)` of the `JuliaSet` class
- fix the output size image (in inches)  
- `README.md`

## [0.2.0] - 2021-07-31

### Added
- `CHANGELOG.md`
- `mirror` option to the `run()` method
- `julia-output-mirror-example.png`
- `.github/workflows/python3-8.yaml` to test Github-actions
- `status badge` for the `README.md` file.
    
### Changed
- normalization is now an option `norm` of the `run()` method
- general improvement of the `docstring` contents

### Fixed

## [0.1.1] - 2021-07-31

### Added

### Changed

### Fixed
- `requirements.txt` for `pip`

## [0.1.0] - 2021-07-31
- Initial commit

tests/test_julia.py
import unittest
import random
import os
from juliaset import julia

class TestJuliaSet(unittest.TestCase):

    def test_julia_default_param(self):

        # Initialize a JuliaSet instance
        newset = julia()

        self.assertEqual(newset.size, 256)
        self.assertEqual(newset.dpi, 300)
        self.assertEqual(newset.norm, True)
        self.assertEqual(newset.mirror, False)
        self.assertEqual(newset.escrad, 3)
        self.assertEqual(newset.niter, 250)

    def test_julia_custom_param(self):

        # Initialize a JuliaSet instance
        newset = julia(size=128, dpi=600, norm=False, mirror=True,
            escrad=4, niter=500)

        self.assertEqual(newset.size, 128)
        self.assertEqual(newset.dpi, 600)
        self.assertEqual(newset.norm, False)
        self.assertEqual(newset.mirror, True)
        self.assertEqual(newset.escrad, 4)
        self.assertEqual(newset.niter, 500)

tests/test_class_juliaset.py
import unittest
import random
import os
from juliaset import JuliaSet

class TestJuliaSet(unittest.TestCase):

    def test_initialize(self):

        # Initialize a JuliaSet instance
        juliaInstance = JuliaSet()

        self.assertEqual(juliaInstance.size, 256)
        self.assertEqual(juliaInstance.dpi, 300)
        self.assertEqual(juliaInstance.norm, True)
        self.assertEqual(juliaInstance.mirror, False)
        self.assertEqual(juliaInstance.escrad, 3)
        self.assertEqual(juliaInstance.niter, 250)

    def test_param_valid_keys(self):

        # Initialize a JuliaSet instance
        juliaInstance = JuliaSet()

        # Call param function
        juliaInstance.param(
            size=128, dpi=600, norm=False, mirror=True,
            escrad=4, niter=500
        )

        self.assertEqual(juliaInstance.size, 128)
        self.assertEqual(juliaInstance.dpi, 600)
        self.assertEqual(juliaInstance.norm, False)
        self.assertEqual(juliaInstance.mirror, True)
        self.assertEqual(juliaInstance.escrad, 4)
        self.assertEqual(juliaInstance.niter, 500)

    def test_twearkComplex(self):

        # Initialize a JuliaSet instance
        juliaInstance = JuliaSet()

        cpxTmp = (1., 1.)
        cpxNum = juliaInstance.twearkComplex(cpxTmp)

        self.assertTrue(cpxNum.real >= 0.98 and cpxNum.real <= 1.02)
        self.assertTrue(abs(cpxNum.imag) >= 0.98 and abs(cpxNum.imag) <= 1.02 )
    
    def test_getComplexValue(self):

        # Initialize random seed
        random.seed(0)

        # Initialize a JuliaSet instance
        juliaInstance = JuliaSet()

        cpxNum = juliaInstance.getComplexValue()

        self.assertAlmostEqual(cpxNum, complex(0.36319, -0.101861), delta=1.e-5)

    def test_getTargetArea(self):

        # Initialize a JuliaSet instance
        juliaInstance = JuliaSet()

        xrng, yrng = juliaInstance.getTargetArea()

        self.assertTrue(xrng[0] >= -1.5 and xrng[0] <= 1.0)
        self.assertTrue(xrng[1] >= -1.0 and xrng[1] <= 1.5)
        self.assertTrue(yrng[0] >= -1.5 and yrng[0] <= 1.0)
        self.assertTrue(yrng[1] >= -1.0 and yrng[1] <= 1.5)
        
        self.assertTrue(xrng[1]-xrng[0] == 1.)
        self.assertTrue(yrng[1]-yrng[0] == 1.)

    def test_processJulia(self):

        # Initialize random seed
        random.seed(0)

        # Initialize a JuliaSet instance
        juliaInstance = JuliaSet()

        # Call param function and fix size at 8 for small test
        juliaInstance.param(
            size=8, dpi=600, norm=False, mirror=False,
            escrad=4, niter=500
        )

        # Fix cpxNum, xrng and yrng
        cpxNum = complex(1.0, 1.0)
        xrng = (-0.5, 0.5)
        yrng = (-0.5, 0.5)

        # call the function
        julia = juliaInstance.processJulia(cpxNum, xrng, yrng)

        attemptOutput = [
            [-0.91055727, -0.91055727, -0.9225403, -0.91055727, -0.91055727, -0.91055727, -0.91055727, -0.91055727],
            [-0.91055727, -0.91055727, -0.91055727, -0.91055727, -0.91055727, -0.91055727, -0.91055727, -0.91055727],
            [-0.91055727, -0.91055727, -0.91055727, -0.91055727, -0.91055727, -0.91055727, -0.91055727, -0.91055727],
            [-0.91055727, -0.91055727, -0.91055727, -0.91055727, -0.91055727, -0.91055727, -0.91055727, -0.91055727],
            [-0.91055727, -0.91055727, -0.91055727, -0.91055727, -0.91055727, -0.91055727, -0.91055727, -0.91055727],
            [-0.91055727, -0.91055727, -0.91055727, -0.91055727, -0.91055727, -0.91055727, -0.91055727, -0.91055727],
            [-0.91055727, -0.91055727, -0.91055727, -0.91055727, -0.91055727, -0.91055727, -0.91055727, -0.91055727],
            [-0.91055727, -0.91055727, -0.91055727, -0.91055727, -0.91055727, -0.91055727, -0.91055727, -0.91055727]
        ]

        for i in range(8):
            for j in range(8):
                self.assertAlmostEqual(julia[i][j], attemptOutput[i][j], delta=1.e-5)
        

juliaset/juliaset.py
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.cm as cm
import random

class JuliaSet:

    def __init__(self):
        """
        Constructor of the JuliaSet class

        :param size: size in pixels (for both width and height)
        :param dpi: dots per inch (default 300)
        """

        # Initialize image related parameters
        self.size = 256
        self.dpi = 300
        self.norm = True
        self.mirror = False

        # Initialize process related parameters
        self.escrad = 3
        self.niter = 250

    def param(self, **kwargs):
        """
        Get parameters from input dictionary and set attributes.

        :param kwargs: a dictionary in the form 
            `{'arg1':value, ..., 'argN': value}`
        """

        # Check if kwargs in not empty
        if kwargs is not None:
            # Image related parameters
            if 'size' in kwargs:
                self.size = kwargs.pop('size', 256)
            if 'dpi' in kwargs:
                self.dpi = kwargs.pop('dpi', 300)
            if 'norm' in kwargs:
                self.norm = kwargs.pop('norm', True)
            if 'mirror' in kwargs:
                self.mirror = kwargs.pop('mirror', False)

            # Process related parameters
            if 'escrad' in kwargs:
                self.escrad = kwargs.pop('escrad', 3)
            if 'niter' in kwargs:
                self.niter = kwargs.pop('niter', 250)

        # If kwargs is not empty there is some invalid keywords
        if kwargs:
            print("{} are invalid keyword arguments!".format(kwargs.keys()))

    def run(self, show=False, fname='juilaset-output'):
        """
        Run the Julia set generator

        :param mirror: if True the julia is mirrored horizontally and
            vertically; each mirror is concatenate with the original
            to produce a new image
        :param norm: if true the Julia set is normalized by its
            absolute maximum value.
        :param show: if show is `False` th eoutput image will be
            written as a PNG file named `fname`
        :param fname: Name of the output PNG file to write on disk
        """

        # Get a complex value among a list of best Julia sets 
        cpxNum = self.getComplexValue()

        # Get the target area
        # For more randomness, the target area is a random 
        # subset of a wide one defined with x[-1.5, 1.5] and
        # y[-1.5, 1.5]
        xrng, yrng = self.getTargetArea()

        # Process
        julia = self.processJulia(cpxNum, xrng, yrng)

        # Normalization
        if(self.norm):
            julia /= np.amax(np.abs(julia))

        # Mirroring
        if(self.mirror):
            # Horizontal mirroring and concatenate
            juliamirror = np.flip(julia, axis=1)
            julia = np.concatenate((julia, juliamirror), axis=1)
            # Vertical mirroring and concatenate
            juliamirror = np.flip(julia, axis=0)
            julia = np.concatenate((julia, juliamirror), axis=0)

        # Plot the output with a random colormap using matplotlib
        self.plotJuliaSet(julia, show=show, fname=fname)

    def getComplexValue(self):
        """
        Random choice in a list of best complex values for Julia 
        sets (real, imag).

        :return cpxNum: a semi-random complex value
        """

        # Define the list of best complex values 
        cpxList = [
            (-0.10, 0.650), (0.00, 0.80), (0.370, 0.100), 
            (0.355, 0.355), (-0.54, 0.54), (0.340, -0.05),
            (0.37, 0.10), (0.355, 0.355)
        ]

        # Randomly choose one
        cpxTmp =  random.choice(cpxList)

        # Manipulate the base value slightly to make it a little more unique
        cpxNum = self.twearkComplex(cpxTmp)

        return cpxNum

    def twearkComplex(self, cpxTmp):
        """
        Manipulate the base value slightly to make it a little more unique.

        :param cpxTmp: complex value to modify
        :param cpxNum: a slightly manipulate version of the input
        """

        # Get the signs for the imaginary parts
        isign = random.randrange(-1, 1, 2)

        # Get a value variation for for real and imaginary parts
        # The possible variation range is fixed at +/- 2% to stay
        # In the neightborhood of the initial value
        rsigma = random.uniform(0.98, 1.02)
        isigma = random.uniform(0.98, 1.02)

        # Apply modification and return the new complex value
        realPart = cpxTmp[0] * rsigma
        imagPart = cpxTmp[1] * isigma * isign

        return complex(realPart, imagPart)

    def getTargetArea(self):
        """
        For more randomness, the target area is a random 
        subset of a wide one defined with x[-1.5, 1.5] and
        y[-1.5, 1.5]

        :return xrng, yrng: tuples containing (xmin, xmax)
            and (ymin, ymax)
        """

        # Randomly choose the center of the target area
        # Possible values are in [-1.0, 1.0] to stay in an
        # area where there are always pieces of fractals
        xctr = random.uniform(-1.0,1.0)
        yctr = random.uniform(-1.0,1.0)

        # Extend around the center
        xrng = (xctr-0.5, xctr+0.5)
        yrng = (yctr-0.5, yctr+0.5)

        return xrng, yrng

    def processJulia(self, cpxNum, xrng, yrng):
        """
        Calculate the Julia set for the given input parameters.

        :param cpxNum: complex value acting as a seed for the Julia set
        :param xrng: range of values (min, max) for the x-axis
        :param yrng: range of values (min, max) for the y-axis
        :param escrad: escape radius
        :param niter: maximum number of iterations
        """

        # Initialize numpy array of dimensions (size, size) with zeros
        julia = np.ones((self.size, self.size), dtype=np.float32)

        # Calculate the width (equal to height) of the image since the
        # image is defined as a square
        width = xrng[1] - xrng[0]  # xmax - xmin = ymax - ymin

        # Randomly choose the sign of the shade
        #ssign = random.randrange(-1, 1, 2)
        ssign = -1.

        # Loop over x range
        for ix in range(self.size):
            # Get the pixel position in the complex plane
            # For the real part
            realPart = float(ix) / self.size * width + xrng[0]

            # Loop over y range
            for iy in range(self.size):
                # Get the pixel position in the complex plane
                # For the imaginary part
                imagPart = float(iy) / self.size * width + yrng[0]

                # Build the complex
                cpxTmp = complex(realPart, imagPart)

                # Initialize iteration counter
                it = 0

                # Loop over iterations
                while(np.abs(cpxTmp) <= self.escrad**2 and it < self.niter):
                    # Quadratic polynomial
                    cpxTmp = cpxTmp**2 + cpxNum
                    # Increment iteration counter
                    it += 1
                
                # Calculate the shade (a cool thing find somewhere on the net)
                shade = 1. - np.sqrt(it/self.niter)

                # Fill the outpout array
                julia[ix][iy] = ssign * shade

        return julia

    def plotJuliaSet(self, julia, fname='juilaset-output', show=False):
        """
        Plot the output Julia set and show it in matplotlib window or
        write it on disk as a png file.

        :param julia: the Julia set
        :param show: if show is `False` th eoutput image will be
            written as a PNG file named `fname`
        :param fname: Name of the output PNG file to write on disk
        """

        # List of beautiful colormap for Julia sets
        cmapList = [
            cm.Blues, cm.Greens, cm.Purples, cm.hot, cm.inferno, 
            cm.binary, cm.rainbow, cm.twilight_shifted, cm.plasma
        ]

        # Randomly chose one colormap
        cmapName = random.choice(cmapList)

        # Plot the image with a gaussian interpolation
        fig = plt.gcf()
        fig.set_size_inches(3., 3.)
        plt.imshow(julia, interpolation='gaussian', cmap=cmapName)
        
        # Disable axis
        plt.axis('off')

        if(show):
            plt.show()
        else:
            # Write on disk
            fig.savefig(fname+".png", dpi=self.dpi, pad_inches=0.05, bbox_inches='tight')

def julia(**kwargs):
    """
    temp
    """
    # Initialize Julia Set instance
    juliaInstance = JuliaSet()

    # If kwargs not empty update the attributes
    if kwargs is not None:
        juliaInstance.param(**kwargs)

    return juliaInstance

if __name__ == "__main__":
    # execute only if run as a script
    genJuliaSet = JuliaSet()
    genJuliaSet.param()
    genJuliaSet.run()
